spec/unit/etc_spec.rb

```diff
@@ -1,234 +1,325 @@
+
+  let(:utf_8_user_struct) do
+    user = Etc::Passwd.new
+    # user name contains same EUC_KR bytes labeled as UTF-8
+    user.name = euc_kr_as_utf_8
+    # group passwd field is valid UTF-8
+    user.passwd = mixed_utf_8
+    user
+  end
+
+  let(:euc_kr_user_struct) do
+    user = Etc::Passwd.new
+    user.name = euc_kr
+    user.passwd = mixed_utf_8_as_euc_kr
+    user
+  end
+
+  let(:ascii_user_struct) do
+    user = Etc::Passwd.new
+    user.name = euc_kr_as_binary
+    user.passwd = mixed_utf_8_as_binary
+    user
+  end
+
+  shared_examples "methods that return an overridden group struct from Etc" do |params|
+    context "when Encoding.default_external is UTF-8" do
+      before do
+        Etc.expects(subject).with(*params).returns(utf_8_group_struct)
+      end
+
+      let(:overridden) {
+        PuppetSpec::CharacterEncoding.with_external_encoding(Encoding::UTF_8) do
+          Puppet::Etc.send(subject, *params)
         end
+      }
+
+      it "should leave the valid UTF-8 values in arrays unmodified" do
+        expect(overridden.mem[0]).to eq(mixed_utf_8)
+        expect(overridden.mem[1]).to eq(root)
+      end
+
+      it "should leave the invalid UTF-8 values in arrays unmodified"do
+        expect(overridden.mem[2]).to eq(euc_kr_as_utf_8)
+      end
+
+      it "should leave the valid UTF-8 values unmodified" do
+        expect(overridden.passwd).to eq(mixed_utf_8)
       end
 
-      it "should return a struct with a :mem array with all field values converted to UTF-8" do
-        converted.mem.each { |elem| expect(elem.encoding).to eq(Encoding::UTF_8) }
+      it "should leave the invalid UTF-8 values unmodified" do
+        expect(overridden.name).to eq(euc_kr_as_utf_8)
       end
     end
 
-    context "given an original Etc::Group struct with field values that cannot be converted to UTF-8" do
-      let(:group) { Etc::Group.new }
+    context "when Encoding.default_external is EUC_KR (i.e., neither UTF-8 nor POSIX)" do
       before do
-        # group membership contains valid and invalid UTF-8
-        group.mem = [convertible_binary, not_convertible]
-        # group name contains a value that is invalid UTF-8
-        group.name = not_convertible
-        # group passwd field is valid UTF-8
-        group.passwd = convertible_binary
-
-        Etc.expects(:getgrent).returns(group)
+        Etc.expects(subject).with(*params).returns(euc_kr_group_struct)
       end
 
-      let(:converted) { Puppet::Etc.getgrent }
+      let(:overridden) {
+        PuppetSpec::CharacterEncoding.with_external_encoding(Encoding::EUC_KR) do
+          Puppet::Etc.send(subject, *params)
+        end
+      }
 
-      it "should convert convertible values in arrays to UTF-8" do
-        expect(converted.mem[0]).to eq("A\u06FF\u16A0\u{2070E}")
-        expect(converted.mem[0].encoding).to eq(Encoding::UTF_8) # just being explicit
+      it "should override EUC_KR-labeled values in arrays to UTF-8 if that would result in valid UTF-8" do
```

Comment:
These were invalid, and they were being masked because we were modifying the expected original values. Pushed a new commit that ensures we're acting on copies for our tests so that the originals go unmodified.